### PR TITLE
donation pages

### PR DESCRIPTION
**Pull request by turino**
I think everything is up to spec. Not sure if the controller responses/views are exactly what's wanted, but it's a start.

to do
- [x] more validations for DonationPage model?
- [x] case insensitive unique index on slug in db (test passes on local machine but fails on Travis)
- [x] clean up tests

### CODE REVIEW

File: spec/models/donation_page_spec.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      best_page = create(:donation_page)
+      ok_page = create(:donation_page)
+      create(:action, donation_page: worst_page, donation_amount_in_cents: 100)
+      create(:action, donation_page: ok_page, donation_amount_in_cents: 300)
+      create(:action, donation_page: best_page, donation_amount_in_cents: 200)
+      create(:action, donation_page: best_page, donation_amount_in_cents: 200)
+
+      results = DonationPage.by_funds_raised
+
+      expect(results).to eq [best_page, ok_page, worst_page]
```
Comment (turino) [inline spec/models/donation_page_spec.rb]: I tried running `expect(results.count).to eq 3` and I got a weird postgres error.

```
SELECT COUNT(donation_pages.id, donation_pages.slug, donation_pages.visible_user_name, sum(actions.donation_amount_in_cents) as funds_raised_in_cents) AS count_donation_pages_id_donation_pages_slug_donation_pages_visi, donation_pages.id, donation_pages.slug, donation_pages.visible_user_name AS donation_pages_id_donation_pages_slug_donation_pages_visible_us FROM "donation_pages" INNER JOIN "actions" ON "actions"."donation_page_id" = "donation_pages"."id" GROUP BY donation_pages.id, donation_pages.slug, donation_pages.visible_user_name  ORDER BY funds_raised_in_cents desc
PG::SyntaxError: ERROR:  syntax error at or near "as"
LINE 1: ..._user_name, sum(actions.donation_amount_in_cents) as funds_r...
                                                             ^
```

Does it matter if we can't call `count` on this scope?

Comment (Rio517) [inline spec/models/donation_page_spec.rb]: Yeah, AREL is wrapping your select statement in `COUNT()`.  Did you see the crazy `as `count_donation_pages_id_donation_pages_slug_donation_pages_visi`?  We can just use length where neccessary.  It occured to me, a count on the number of donations would be really useful. I added that to your scope method.  Thanks for all this work!

**Comment by Rio517** [comment]
Hi, 

This is almost prefect.

Can we add paperclip uploads to the create and update request?  I won't just be sending a URL I also need to handle sending to s3?  Let me know if you have time to help!

Thanks!

-Mario

**Comment by Rio517** [comment]
Hum..  

I like having tests as close as possible to the code they're testing and I believe in testing liberally.  If we bubble something up only into the request spec, it becomes less clear where the failure is.  I think these tests are good.  Also, by using subject, we're only hitting the DB in subject - 1x not 4x.

There is always a balance. I place a high premium on clarity and development speed.  The later usually equating to test speed. Having been burned by suites that take too long to run (30 minutes one one project), I try to be thoughtful about using let, before and subject to keep execution DRY.  I think this is a good candidate.

In cases like these, where were dealing with a object (our results) with multiple points of failure, I like breaking out specs into multiple.  If a bug develops you can see what failed more quickly.  This case is a little more simple, but I also don't like having 4 implicit expectations in one test.  I try to keep a match between my test name and the given expectation.

What is upcases's/thoughtbot's take on let, subject and before?  a link you can reference?  I'd guess they encourage using doubles and spies.  ??

**Comment by turino** [comment]
The `subject` block runs for each `it`. (I just tested with byebug)

Thoughtbot says to avoid subject, let, before, etc: https://github.com/thoughtbot/guides/tree/dee16052f49cb4b41fb8d090ffdb75942512ddb1/best-practices#testing

Here's the weekly iteration that got me thinking about this: https://upcase.com/videos/rspec-best-practices
The issues they're talking about are not a big problem with this specific example because `subject` is close to its `it` blocks, though, so it may not be a big deal.

If this test didn't touch the database I'd agree that it's better to put each expectation in its own spec, but since we are creating several records for each test I'm not so sure. Also, I think the expectations *were* explicit, they just didn't have their own test name. You get the same error message either way, you just don't get the helpful "correctly calculates funds_raised_in_cents". I guess you also don't get to see if you have multiple failures...

**Comment by turino** [comment]
@Rio517 I'm ready to push my rebase, but wanted to be sure we're done with this discussion before I do.  If you want to keep the separate expectations, can we at least break the `subject` into `before` and `subject`? It will do the same thing, but I think will be clearer.

Also, do you think `expect(subject.length).to eq(3)` is really necessary?

**Comment by Rio517** [comment]
@turino If we break subject into before and subject, I'm pretty sure it will require the 4 database calls you were referring to.  We could do before(:all), but i prefer not to go that route.

Yeah, good flag on removing the length.  I was defensively coding against myself, so that in the future it would remind me to use length instead of count, but I'd probably figure that out before I ever saw the spec.

On a general note, since MAYDAY is a non-profit, and I don't know who is going to be maintaining this in the future, I tend to code more defensively than I might in other projects and go out of my way to be extra explicit.

**Comment by turino** [comment]
Did you see my earlier message? The code in subject runs for each 'it'
block. I tested it with byebug. I think before(:all) is the only way to
avoid it. I also prefer not to go that route

On Thursday, January 14, 2016, Mario Olivio Flores <notifications@github.com>
wrote:

> @turino <https://github.com/turino> If we break subject into before and
> subject, I'm pretty sure it will require the 4 database calls you were
> referring to. We could do before(:all), but i prefer not to go that route.
>
> Yeah, good flag on removing the length. I was defensively coding against
> myself, so that in the future it would remind me to use length instead of
> count, but I'd probably figure that out before I ever saw the spec.
>
> On a general note, since MAYDAY is a non-profit, and I don't know who is
> going to be maintaining this in the future, I tend to code more defensively
> than I might in other projects and go out of my way to be extra explicit.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MayOneUS/mayday-2.0-backend/pull/74#issuecomment-171683229>
> .
>


-- 
sent from phone

**Comment by Rio517** [comment]
@turino Yeah, I verified that subject does rerun the execution on every example. I had hoped it didn't.  I like have something to explicitly pass off my what's being tested to the test example, instead of setting a variable in before and I even perfer it to `let` in the sense that subject tells you that this is the one thing you are testing against.  In the specific case we're talking about, I'd argue our examples should be broken out since they're validating different things.  Let's discuss on the phone in more detail, perhaps.

I still kept getting a failing test due to database cleaning issues. I updated the rspec helper to clean around examples instead of a before/after call. Seems to fix my local issue, but I'm still seeing this uniquness validation fail in TravisCI. I was about to take a look at the link you referenced in the spec file.

**Comment by Rio517** [comment]
Thing I think are pending:
- [X] reciving a url string is fine - no additional work needed there.
- [ ] add optional field - goal in cents - a dollar goal in cents.
- [ ] update json response from get to include total count of donations and total funds raised - see old planning document for updated example of json response (authoritative reference)
- [ ] when do we do edit, can't use passed in slug - identify based on uuid of donation page. 
- [ ]  create response, should return a uuid

Let's worry about reuqesting auth tokens to edit the pages later.

**Pull request merged**